test/aem/aemupload.test.js

```diff
@@ -284,4 +284,209 @@ describe('AEM Upload', function() {
+    });
+
+    it('AEM upload special characters in target path and asset name', async function() {
+        await setupAemUploadTest('/path/with # special characters','file-1 #.jpg', true);
+        await setupAemUploadTest('/path/with & special characters/included','file-1 &.jpg', true);
+        await setupAemUploadTest('/path/with + special characters','file-1 +.jpg', true);
+        await setupAemUploadTest('/path/with ( special characters','file-1 (.jpg', true);
+        await setupAemUploadTest('/path/with ) special characters/included','file-1 ).jpg', true);
+        await setupAemUploadTest('/path/with # special & characters (+)','file-1 # & + ( ).jpg', true);
+    });
+
+    it('AEM upload special characters target path and asset name already encoded', async function() {
+        await setupAemUploadTest('/path/with%20%23%20special%20characters','file-1%20%23.jpg', true);
+    });
+
+    it('AEM upload special characters unicode in target path and asset name', async function() {
+        await setupAemUploadTest('/path/ഘങചഛ/থদধনপ','໓໔໕໖.jpg', true);
+    });
+
+    it('AEM upload special characters in asset name, special character support disabled', async function() {
+        const testFile = Path.join(__dirname, 'file-1.jpg');
+        await fs.writeFile(testFile, 'hello world 123', 'utf8');
+        const aemUpload = new AEMUpload();
+        
+        const events = {
+            filestart: [],
+            fileprogress:[],
+            fileend: [],
+            fileerror: []
+        };
+        aemUpload.on('fileerror', (data) => {
+            events.fileerror.push(data);
+        });
+        await aemUpload.uploadFiles({
+            uploadFiles: [{
+                fileUrl: `http://test-aem-upload-201/path/to/file-1&.jpg`,
+                filePath: testFile,
+                fileSize: 15
+            }],
+            supportSpecialCharacters: false
+        });
+
+        const errors = events.fileerror[0].errors;
+        assert.strictEqual(errors.length, 1);
+        assert.strictEqual(errors[0].message, "File cannot be uploaded: Filename 'file-1&.jpg' has unsupported characters");
+    });
 });
+
+/**
+ * Mocks a simple AEMUpload and verifies upload event data.
+ *
+ * @param {String} assetPath path to mock asset
+ * @param {String} assetName name of a mock asset
+ * @param {boolean} supportSpecialCharacters if true, special characters will be encoded
+ *  during the test. Otherwise no encoded will be performed.
+ */
+async function setupAemUploadTest(assetPath, assetName, supportSpecialCharacters = false) {
+    
+    /**
+     * Simulates HTML form encoding (application/x-www-form-urlencoded).
+     *
+     * @param {String} value to be encoded
+     * @returns {String} encoded value.
+     */
+    function getFormEncoding(value) {
+        const form = new URLSearchParams();
+        form.append('myKey', decodeURIComponent(value));
```

Comment:
Why is this decoding the value?

Reply:
This value is being decoded to support a [unit test where the target path is already URL encoded](https://github.com/adobe/node-httptransfer/pull/130/files#diff-42bb1c8fc4d7014d2ca5b6b5e77959d431bc7e600f27295d81d60b9f58a61476R306). The decoding here in the test code matches decoding performed on [target paths and filenames](https://github.com/adobe/node-httptransfer/blob/17007fe84fc6d933b0849b9502a3b09cd69d3131/lib/util.js#L203) when `TransferAsset` [eventData is requested](https://github.com/adobe/node-httptransfer/blob/17007fe84fc6d933b0849b9502a3b09cd69d3131/lib/asset/transferasset.js#L201).